py/dml/traits.py

```diff
@@ -559,7 +559,8 @@ def mark_referenced(self, trait):
         if not sub in self.referenced:
             self.referenced.add(sub)
             for m in list(self.method_overrides.values()):
-                codegen.mark_method_referenced(codegen.method_instance(m))
+                if sub.member_declaration(m.name) is not None:
+                    codegen.mark_method_referenced(codegen.method_instance(m))
```

Comment:
Sidenote: iterating through all method overrides of the node for every trait marked referenced strikes me as very inefficient.